auth/create-auth-tables.sql

```diff
@@ -1,22 +1,23 @@
-CREATE TABLE `user_data` (
+CREATE TABLE `users` (
   `id` INT(11) NOT NULL AUTO_INCREMENT,
+  `state` VARCHAR(100) NOT NULL,
+  -- creating, active, deleting, deleted
   `username` varchar(255) NOT NULL,
-  `user_id` varchar(255) NOT NULL,
   `email` varchar(255) DEFAULT NULL,
-  `developer` tinyint(1) DEFAULT NULL,
-  `service_account` tinyint(1) DEFAULT NULL,
+  `is_developer` tinyint(1) DEFAULT NULL,
```

Comment:
Above email should be not null, the unique constraint on email will break after the first null entry

Reply:
That's not how unique key constraints work.  NULL always compares false, and two rows are duplicate if their keys are equal, so you can have as many unique NULL values as you want.  For example:

```
mysql> CREATE TABLE `foo` (
    ->   `email` varchar(255) DEFAULT NULL,
    ->   UNIQUE KEY `email` (`email`)
    -> ) ENGINE=InnoDB;
Query OK, 0 rows affected (0.02 sec)

mysql> INSERT INTO foo(email) VALUES (NULL);
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO foo(email) VALUES (NULL);
Query OK, 1 row affected (0.00 sec)

mysql> select * from foo;
+-------+
| email |
+-------+
| NULL  |
| NULL  |
+-------+
2 rows in set (0.00 sec)
```

I want email to be NULL for service accounts that have no email and cannot log in.

Reply:
It looks like this is SQL implementation dependent: https://dba.stackexchange.com/questions/80514/why-does-a-unique-constraint-allow-only-one-null
https://stackoverflow.com/questions/9280680/stop-mysql-tolerating-multiple-nulls-in-a-unique-constraint

Interesting: `NULL = 1` evaluate to NULL (as it does for any other arithmetic operator), so the unique key check must be using the boolean operator IS (which will evaluates NULL IS NULL to 0).
  - https://dev.mysql.com/doc/refman/8.0/en/working-with-null.html 

MySQL tolerates it, thanks for the clarification!